app/services/base_services/set_attributes.rb

```diff
@@ -35,6 +35,13 @@ class SetAttributes
     def initialize(user:, model:, contract_class:)
       self.user = user
       self.model = model
+
+      # Allow tracking changes caused by a user but done for him by the system.
+      # E.g. fixed_version of a work package might need to be changed as the user changed the project.
+      # This is currently used for permission checks where the changed project is checked but the fixed_version
+      # is not if it is done by the system.
+      model.extend(Mixins::ChangedBySystem)
```

Comment:
I don't like this pattern as it will make it very hard to reason about where this functionality has been included. Magically code returned by the service will have more methods. It's fine now, but this is a source for heisenbugs as soon as `ChangedBySystem` introduces any (even private) method that overrides something else Why not put this in the ApplicationRecord base class instead?

Reply:
The dislike is understandable, I share it but currently do not have a better solution. As it is rather specific to the interaction between contract and service however, I wouldn't want to place it in the 'global' AR base class.